File: src/main/java/dp/LC5LongestPalindromicSubstring.java
package dp;

public class LC5LongestPalindromicSubstring {
//    Given a string s, find the longest palindromic substring in s. You may assume that the maximum length of s is 1000.
//
//    Example 1:
//
//    Input: "babad"
//    Output: "bab"
//    Note: "aba" is also a valid answer.
//            Example 2:
//
//    Input: "cbbd"
//    Output: "bb"
    //Manacher's Algorithm

    public String longestPalindrome(String s) {
        int n = 2 * s.length() + 1;
        char ch[] = new char[n];
        for (int i = 0; i < n - 1; i += 2) { //construct a new string by insert # between every two adjacent characters
            ch[i] = '#';
            ch[i + 1] = s.charAt(i / 2);
        }
        ch[n - 1] = '#';
        int arr[] = new int[n];  //Palindrome radius
        int R = 0, RC = 0; //RC is Palindrome center, R is Palindrome right boundary
        for (int i = 1; i < n; i++) {
            if (i >= R) {
                int j = 1;
                for (; i - j >= 0 && i + j < n && ch[i - j] == ch[i + j]; j++) ;
                R = i + j - 1;
                RC = i;
                arr[i] = j - 1;
            } else {
                int iL = RC - (i - RC);
                if (iL - arr[iL] < RC - arr[RC])
                    arr[i] = R - i;
                else if (iL - arr[iL] > RC - arr[RC])
                    arr[i] = arr[iL];
                else {
                    int j = arr[iL];
                    for (; i - j >= 0 && i + j < n && ch[i - j] == ch[i + j]; j++) ;
                    R = i + j - 1;
                    RC = i;
                    arr[i] = j - 1;
                }
            }
        }

        int maxR = 0, maxRC = 0;
        for (int i = 0; i < n; i++) {
            if (arr[i] > maxR) {
                maxR = arr[i];
                maxRC = i;
            }
        }
        return s.substring((maxRC - maxR) / 2, (maxRC + maxR) / 2);
    }
}

File: src/main/scala/dp/LC5_LongestPalindromicSubstring.scala
package dp

import scala.annotation.tailrec


object LC5_LongestPalindromicSubstring {
  //  Given a string s, find the longest palindromic substring in s. You may assume that the maximum length of s is 1000.
  //
  //  Example 1:
  //
  //  Input: "babad"
  //  Output: "bab"
  //  Note: "aba" is also a valid answer.
  //  Example 2:
  //
  //  Input: "cbbd"
  //  Output: "bb"
  def longestPalindrome(s: String): String = {

    @tailrec
    def maxAt(startIndex: Int, endIndex: Int): String = {
      if (startIndex < 0 || endIndex > s.length - 1 || s.charAt(startIndex) != s.charAt(endIndex))
        s.substring(startIndex + 1, endIndex)
      else
        maxAt(startIndex - 1, endIndex + 1)
    }

    @tailrec
    def lp(st: Int, largestYet: String): String = {
      if (st == s.length) largestYet
      else {
        val largestOdd = maxAt(st - 1, st + 1)
        val largestEven = maxAt(st, st + 1)

        if (largestOdd.length > largestEven.length && largestOdd.length > largestYet.length)
          lp(st + 1, largestOdd)
        else if (largestEven.length > largestOdd.length && largestEven.length > largestYet.length)
          lp(st + 1, largestEven)
        else
          lp(st + 1, largestYet)
      }
    }

    lp(0, "")
  }
}
